src/model/data_frame.cpp

```diff
@@ -1,2 +1,22 @@
-
 #include "data_frame.h"
+#include "services/service_manager.h"
+#include "utility/json_utils.h"
+#include "utility/time_utils.h"
+#include <boost/property_tree/ptree.hpp>
+
+record_data_frame_t record_data_frame_t::createDataFrame() {
+    const std::vector<std::string>& keys = SM::getDataCache()->getKeys();
+    const boost::property_tree::ptree cache = SM::getDataCache()->toPropertyTree();
+    auto dataFrame = SM::getDatabase()->createDataFrame({-1,
+                                                         record_system_t::getActiveSystem().id,
+                                                         getCurrentMillis(),
+                                                         propertyTreeToString(cache)});
+
+    for (auto key : keys) {
+        boost::property_tree::ptree data;
+        data.add(key, cache.get_child(key).data());
+        SM::getDatabase()->createDataPoint(
```

Comment:
Can we move this part out of the data frame code? The code to save individual data points is related but independent, so I'd rather have it in the main file or a separate function, below where you create the data frame.